extensions/ql-vscode/src/vscode-tests/test-config.ts

```diff
@@ -0,0 +1,93 @@
+import { ConfigurationTarget, workspace } from 'vscode';
+import { CUSTOM_CODEQL_PATH_SETTING, REMOTE_CONTROLLER_REPO, REMOTE_REPO_LISTS, Setting } from '../config';
+
+class TestSetting<T> {
+  private settingState: {
+    globalValue?: T;
+    workspaceValue?: T,
+    workspaceFolderValue?: T,
+  } | undefined;
+
+  constructor(
+    public readonly setting: Setting,
+    private initialTestValue: T | undefined = undefined
+  ) { }
+
+  public async get(): Promise<T | undefined> {
+    return this.section.get(this.setting.name);
+  }
+
+  public async set(value: T | undefined, target: ConfigurationTarget = ConfigurationTarget.Global): Promise<void> {
+    await this.section.update(this.setting.name, value, target);
+  }
+
+  public async setInitialTestValue(value: T | undefined) {
+    this.initialTestValue = value;
+  }
+
+  public async initialSetup() {
+    this.settingState = this.section.inspect(this.setting.name);
+
+    // Unfortunately it's not well-documented how to check whether we can write to a workspace
+    // configuration. This is the best I could come up with. It only fails for initial test values
+    // which are not undefined.
+    if (this.settingState?.workspaceValue !== undefined) {
+      await this.set(this.initialTestValue, ConfigurationTarget.Workspace);
+    }
+    if (this.settingState?.workspaceFolderValue !== undefined) {
+      await this.set(this.initialTestValue, ConfigurationTarget.WorkspaceFolder);
+    }
+
+    await this.setup();
+  }
+
+  public async setup() {
+    await this.set(this.initialTestValue, ConfigurationTarget.Global);
+  }
+
+  public async restoreToInitialValues() {
+    const state = this.section.inspect(this.setting.name);
+
+    // We need to check the state of the setting before we restore it. This is less important for the global
+    // configuration target, but the workspace/workspace folder configuration might not even exist. If they
+    // don't exist, VSCode will error when trying to write the new value (even if that value is undefined).
+    if (state?.globalValue !== this.settingState?.globalValue) {
+      await this.set(this.settingState?.globalValue, ConfigurationTarget.Global);
+    }
+    if (state?.workspaceValue !== this.settingState?.workspaceValue) {
+      await this.set(this.settingState?.workspaceValue, ConfigurationTarget.Workspace);
+    }
+    if (state?.workspaceFolderValue !== this.settingState?.workspaceFolderValue) {
+      await this.set(this.settingState?.workspaceFolderValue, ConfigurationTarget.WorkspaceFolder);
+    }
+  }
+
+  private get section() {
+    if (this.setting.parent === undefined) {
+      throw new Error('Cannot get the value of a root setting.');
+    }
+    return workspace.getConfiguration(this.setting.parent.qualifiedName);
+  }
+}
+
+export const testConfig = {
```

Comment:
This is going a bit further than the other suggestions, but I wondered if it would be possible to:
- Avoid specifying a list of setting names here:
  - It creates duplication as we have to remember to add any new settings to another location.
  - It also means we don't reset all settings unless we manually write all setting names here.
- Avoid tests having to know about and use the test config:
  - It makes test code different from production code.
  - If some test code calls production code that sets config the normal way, there's a chance this might not get reset. It depends on if that setting is listed here, which it might not be if the test code isn't setting it through the test config.

Instead, could we programatically determine the list of all settings? I think we can. Again I had a go at it just to make sure it was possible. You can see my attempt at https://github.com/github/vscode-codeql/commit/d0eca7e01d61b3191755002abb2cbb120b4b3d9d. That also reverts some changes from this PR that are no longer necessary.

It's worth noting that `contributes.configuration.properties` doesn't include every setting, because putting a setting here makes it public. Therefore some settings such as the canary flag and live results flag are deliberately not in this list. You can still do `getValue` to get that value of such a setting, but you can't do `updateValue` as then vscode complains the setting isn't defined. This means for these setting we have to use [stubs in the tests](https://github.com/github/vscode-codeql/blob/5644206777d6942610189290d6721e20ce3a84f0/extensions/ql-vscode/src/vscode-tests/no-workspace/run-queries.test.ts#L22) but this is fine because stubs are restored after each test anyway.

What do you think?

Reply:
Update...  it doesn't seem to be working and I'm not sure why. Values aren't being reset between tests. I've been using d3d6c4143ae07f3ff9ffc6bda13a9d8acca96ace to test it.

We could have a look into it, either together or separately, if you think it's worth pursuing.

Reply:
I did think about automatically adding all settings. I didn't think about using the `package.json` for it since that doesn't include the canary and live results flags which are the config settings we're now setting most often in tests.

An alternative solution I did think about is keeping track of all created `Setting` objects and using that to keep a list of all used settings. That would mean the extension has some logic in it purely for tests, but this is fairly minimal (it could be as simple as `ALL_SETTINGS.push(this)` in the `Setting` constructor). What do you think about doing it in this way?

Reply:
I think adding a line to the `Setting` constructor that adds it to some list would be ok. I think it would be worth trying it and seeing how well it works in practice.